File: legacy/scroll.lua
--- MouseKeys scroll
--- Time-based ramp while holding, momentum glide on release

local scroll = {}

local config = nil

local pressTime = { up = nil, down = nil, left = nil, right = nil }
local velocity = { x = 0, y = 0 }
local accum = { x = 0, y = 0 }
local tapFired = { x = false, y = false }

local directions = {
  up    = { x = 0,  y = 1  },
  down  = { x = 0,  y = -1 },
  left  = { x = -1, y = 0  },
  right = { x = 1,  y = 0  },
}

function scroll.init(cfg)
  config = cfg
end

function scroll.update(now, dt)
  local anyKeyY = pressTime.up or pressTime.down
  local anyKeyX = pressTime.left or pressTime.right
  local anyKey = anyKeyY or anyKeyX

  local pixelScale = config.scrollPixelScale or 4

  -- While holding: compute speed from hold duration (time-based, no friction)
  -- On release: coast using last velocity with friction decay
  if anyKey then
    local sx, sy = 0, 0
    for dir, startTime in pairs(pressTime) do
      if startTime then
        local holdTime = now - startTime
        local speed = config.scrollInitialSpeed + config.scrollAcceleration * holdTime
        speed = math.min(speed, config.scrollMaxSpeed)
        local vec = directions[dir]
        sx = sx + vec.x * speed
        sy = sy + vec.y * speed
      end
    end
    -- Set velocity for glide on release
    velocity.x = sx
    velocity.y = sy
  else
    -- Glide: friction decay
    local friction = config.scrollFriction
    local frictionDt = friction ^ (dt / 0.016)
    velocity.x = velocity.x * frictionDt
    velocity.y = velocity.y * frictionDt
  end

  -- Tap: instant scroll on first press
  local scrollAmountX, scrollAmountY = 0, 0

  if not tapFired.y and anyKeyY then
    tapFired.y = true
    local tap = (config.scrollTap or 1) * pixelScale
    scrollAmountY = pressTime.up and tap or -tap
  end

  if not tapFired.x and anyKeyX then
    tapFired.x = true
    local tap = (config.scrollTap or 1) * pixelScale
    scrollAmountX = pressTime.right and tap or -tap
  end

  -- Accumulate and extract integer pixels
  accum.x = accum.x + velocity.x * dt * pixelScale
  accum.y = accum.y + velocity.y * dt * pixelScale

  local dx = (accum.x >= 0) and math.floor(accum.x) or math.ceil(accum.x)
  local dy = (accum.y >= 0) and math.floor(accum.y) or math.ceil(accum.y)
  accum.x = accum.x - dx
  accum.y = accum.y - dy

  scrollAmountX = scrollAmountX + dx
  scrollAmountY = scrollAmountY + dy

  if scrollAmountY ~= 0 or scrollAmountX ~= 0 then
    hs.eventtap.scrollWheel({ scrollAmountX, scrollAmountY }, {}, "pixel")
  end

  -- Reset tap state when axis idle
  if not anyKeyY and math.abs(velocity.y) < 0.1 then
    velocity.y = 0
    accum.y = 0
    tapFired.y = false
  end
  if not anyKeyX and math.abs(velocity.x) < 0.1 then
    velocity.x = 0
    accum.x = 0
    tapFired.x = false
  end
end

function scroll.press(dir)
  if not pressTime[dir] then
    pressTime[dir] = hs.timer.absoluteTime() / 1e9
  end
end

function scroll.release(dir)
  pressTime[dir] = nil
end

function scroll.releaseAll()
  for dir in pairs(pressTime) do
    pressTime[dir] = nil
  end
end

function scroll.isIdle()
  for _, startTime in pairs(pressTime) do
    if startTime then return false end
  end
  return math.abs(velocity.x) < 0.1 and math.abs(velocity.y) < 0.1
end

function scroll.stop()
  scroll.releaseAll()
  velocity.x = 0
  velocity.y = 0
  accum.x = 0
  accum.y = 0
  tapFired.x = false
  tapFired.y = false
end

return scroll

File: src/core/scroll.cpp
#include "core/scroll.h"

#include <algorithm>
#include <cmath>

namespace mk {

// Scroll direction vectors: positive y scrolls up (matches the original
// spoon's hs.eventtap.scrollWheel convention).
static const double kVecX[4] = {0, 0, -1, 1};
static const double kVecY[4] = {1, -1, 0, 0};

void Scroll::configure(double initial_speed, double acceleration, double max_speed,
                       double friction, double tap, double pixel_scale) {
  initial_ = initial_speed;
  accel_ = acceleration;
  max_ = max_speed;
  friction_ = friction;
  tap_ = tap;
  pixel_scale_ = pixel_scale;
}

void Scroll::press(Dir d, double now) {
  int i = static_cast<int>(d);
  if (press_time_[i] < 0) press_time_[i] = now;
}

void Scroll::release(Dir d) { press_time_[static_cast<int>(d)] = -1; }

void Scroll::release_all() {
  for (double& t : press_time_) t = -1;
}

void Scroll::reset() {
  release_all();
  vx_ = vy_ = 0;
  ax_ = ay_ = 0;
  tap_fired_x_ = tap_fired_y_ = false;
}

bool Scroll::idle() const {
  for (double t : press_time_)
    if (t >= 0) return false;
  return std::fabs(vx_) < 0.1 && std::fabs(vy_) < 0.1;
}

Delta Scroll::update(double now, double dt) {
  bool any_y = press_time_[0] >= 0 || press_time_[1] >= 0;
  bool any_x = press_time_[2] >= 0 || press_time_[3] >= 0;

  // While holding: speed from hold duration (time-based, no friction).
  // On release: coast on the last velocity with friction decay.
  if (any_x || any_y) {
    double sx = 0, sy = 0;
    for (int i = 0; i < 4; i++) {
      if (press_time_[i] < 0) continue;
      double hold = now - press_time_[i];
      double speed = std::min(initial_ + accel_ * hold, max_);
      sx += kVecX[i] * speed;
      sy += kVecY[i] * speed;
    }
    vx_ = sx;
    vy_ = sy;
  } else {
    double f = std::pow(friction_, dt / 0.016);
    vx_ *= f;
    vy_ *= f;
  }

  // Tap: instant scroll on the first press of an idle axis.
  int out_x = 0, out_y = 0;
  if (!tap_fired_y_ && any_y) {
    tap_fired_y_ = true;
    int tap = static_cast<int>(tap_ * pixel_scale_);
    out_y = press_time_[0] >= 0 ? tap : -tap;
  }
  if (!tap_fired_x_ && any_x) {
    tap_fired_x_ = true;
    int tap = static_cast<int>(tap_ * pixel_scale_);
    out_x = press_time_[3] >= 0 ? tap : -tap;
  }

  ax_ += vx_ * dt * pixel_scale_;
  ay_ += vy_ * dt * pixel_scale_;
  int dx = static_cast<int>(std::trunc(ax_));
  int dy = static_cast<int>(std::trunc(ay_));
  ax_ -= dx;
  ay_ -= dy;
  out_x += dx;
  out_y += dy;

  // Reset tap state once an axis is fully idle.
  if (!any_y && std::fabs(vy_) < 0.1) {
    vy_ = 0;
    ay_ = 0;
    tap_fired_y_ = false;
  }
  if (!any_x && std::fabs(vx_) < 0.1) {
    vx_ = 0;
    ax_ = 0;
    tap_fired_x_ = false;
  }

  return {out_x, out_y};
}

}  // namespace mk
